### ISSUE TITLE
Data race when shutting down terragrunt

### ISSUE THREAD

**Issue by kevinburkeotto** [gruntwork-io/terragrunt#348]
Stack trace here, I hit it when Terragrunt prompted me to enter a value for a variable, and I sent Ctrl+C to the program.

```
WARNING: DATA RACE
Read at 0x00c4203b2a40 by goroutine 40:
  github.com/gruntwork-io/terragrunt/shell.NewSignalsForwarder.func1()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/shell/run_shell_cmd.go:108 +0x1a6

Previous write at 0x00c4203b2a40 by main goroutine:
  os/exec.(*Cmd).Start()
      /Users/kevin/go/src/os/exec/exec.go:363 +0x822
  os/exec.(*Cmd).Run()
      /Users/kevin/go/src/os/exec/exec.go:286 +0x3c
  github.com/gruntwork-io/terragrunt/shell.RunShellCommand()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/shell/run_shell_cmd.go:55 +0x5cb
  github.com/gruntwork-io/terragrunt/shell.RunTerraformCommand()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/shell/run_shell_cmd.go:21 +0x75
  github.com/gruntwork-io/terragrunt/cli.runTerragruntWithConfig()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:263 +0xf4
  github.com/gruntwork-io/terragrunt/cli.runTerragrunt()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:222 +0x130
  github.com/gruntwork-io/terragrunt/cli.runCommand()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:195 +0xb6
  github.com/gruntwork-io/terragrunt/cli.runApp()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:176 +0x1d0
  github.com/gruntwork-io/terragrunt/vendor/github.com/urfave/cli.HandleAction()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/vendor/github.com/urfave/cli/app.go:502 +0xfd
  github.com/gruntwork-io/terragrunt/vendor/github.com/urfave/cli.(*App).Run()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/vendor/github.com/urfave/cli/app.go:268 +0x8fb
  main.main()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/main.go:20 +0x103

Goroutine 40 (running) created at:
  github.com/gruntwork-io/terragrunt/shell.NewSignalsForwarder()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/shell/run_shell_cmd.go:103 +0xb8
  github.com/gruntwork-io/terragrunt/shell.RunShellCommand()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/shell/run_shell_cmd.go:52 +0x589
  github.com/gruntwork-io/terragrunt/shell.RunTerraformCommand()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/shell/run_shell_cmd.go:21 +0x75
  github.com/gruntwork-io/terragrunt/cli.runTerragruntWithConfig()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:263 +0xf4
  github.com/gruntwork-io/terragrunt/cli.runTerragrunt()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:222 +0x130
  github.com/gruntwork-io/terragrunt/cli.runCommand()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:195 +0xb6
  github.com/gruntwork-io/terragrunt/cli.runApp()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:176 +0x1d0
  github.com/gruntwork-io/terragrunt/vendor/github.com/urfave/cli.HandleAction()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/vendor/github.com/urfave/cli/app.go:502 +0xfd
  github.com/gruntwork-io/terragrunt/vendor/github.com/urfave/cli.(*App).Run()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/vendor/github.com/urfave/cli/app.go:268 +0x8fb
  main.main()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/main.go:20 +0x103
```

**Comment by kevinburkeotto** [comment]
I'm not sure which one actually provided the prompt. It's what happens if you declare a variable without a default in vars.tf but don't provide a value for it in terraform.tfvars.

**Comment by kevinburkeotto** [comment]
just hit another one, no prompt, just hit ctrl+c during `terragrunt plan`

```
^C[terragrunt] 2017/11/08 14:33:59 Forward signal interrupt to terraform.
==================
WARNING: DATA RACE
Read at 0x00c4203b2200 by goroutine 41:
  github.com/gruntwork-io/terragrunt/shell.NewSignalsForwarder.func1()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/shell/run_shell_cmd.go:108 +0x1a6

Previous write at 0x00c4203b2200 by main goroutine:
  os/exec.(*Cmd).Start()
      /Users/kevin/go/src/os/exec/exec.go:363 +0x822
  os/exec.(*Cmd).Run()
      /Users/kevin/go/src/os/exec/exec.go:286 +0x3c
  github.com/gruntwork-io/terragrunt/shell.RunShellCommand()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/shell/run_shell_cmd.go:55 +0x5cb
  github.com/gruntwork-io/terragrunt/shell.RunTerraformCommand()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/shell/run_shell_cmd.go:21 +0x75
  github.com/gruntwork-io/terragrunt/cli.runTerragruntWithConfig()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:263 +0xf4
  github.com/gruntwork-io/terragrunt/cli.runTerragrunt()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:222 +0x130
  github.com/gruntwork-io/terragrunt/cli.runCommand()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:195 +0xb6
  github.com/gruntwork-io/terragrunt/cli.runApp()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:176 +0x1d0
  github.com/gruntwork-io/terragrunt/vendor/github.com/urfave/cli.HandleAction()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/vendor/github.com/urfave/cli/app.go:502 +0xfd
  github.com/gruntwork-io/terragrunt/vendor/github.com/urfave/cli.(*App).Run()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/vendor/github.com/urfave/cli/app.go:268 +0x8fb
  main.main()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/main.go:20 +0x103

Goroutine 41 (running) created at:
  github.com/gruntwork-io/terragrunt/shell.NewSignalsForwarder()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/shell/run_shell_cmd.go:103 +0xb8
  github.com/gruntwork-io/terragrunt/shell.RunShellCommand()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/shell/run_shell_cmd.go:52 +0x589
  github.com/gruntwork-io/terragrunt/shell.RunTerraformCommand()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/shell/run_shell_cmd.go:21 +0x75
  github.com/gruntwork-io/terragrunt/cli.runTerragruntWithConfig()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:263 +0xf4
  github.com/gruntwork-io/terragrunt/cli.runTerragrunt()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:222 +0x130
  github.com/gruntwork-io/terragrunt/cli.runCommand()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:195 +0xb6
  github.com/gruntwork-io/terragrunt/cli.runApp()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/cli/cli_app.go:176 +0x1d0
  github.com/gruntwork-io/terragrunt/vendor/github.com/urfave/cli.HandleAction()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/vendor/github.com/urfave/cli/app.go:502 +0xfd
  github.com/gruntwork-io/terragrunt/vendor/github.com/urfave/cli.(*App).Run()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/vendor/github.com/urfave/cli/app.go:268 +0x8fb
  main.main()
      /Users/kevin/src/github.com/gruntwork-io/terragrunt/main.go:20 +0x103
==================
```

**Comment by kevinburkeotto** [comment]
guessing that the problem (and solution) are the same as discussed here https://github.com/99designs/aws-vault/pull/162

**Comment by brikis98** [comment]
Ah, interesting. Does calling `cmd.Start()` guarantee `cmd.Process` will be set?

**Comment by brikis98** [comment]
Ah, looking at the Go src, it seems that it does. In that case, using `Start()` and `Wait()` does seem like the proper fix here. PR welcome!

**Comment by brikis98** [comment]
Fixed by #361.
